ui/ducks/send/send.js

```diff
@@ -0,0 +1,1346 @@
+  KNOWN_RECIPIENT_ADDRESS_WARNING,
+  MIN_GAS_LIMIT_HEX,
+  NEGATIVE_ETH_ERROR,
+} from '../../pages/send/send.constants';
+
+import {
+  addGasBuffer,
+  calcGasTotal,
+  generateTokenTransferData,
+  isBalanceSufficient,
+  isTokenBalanceSufficient,
+} from '../../pages/send/send.utils';
+import {
+  getAdvancedInlineGasShown,
+  getCurrentChainId,
+  getGasPriceInHexWei,
+  getIsMainnet,
+  getSelectedAddress,
+  getTargetAccount,
+} from '../../selectors';
+import {
+  displayWarning,
+  estimateGas,
+  hideLoadingIndication,
+  showConfTxPage,
+  updateTransaction,
+} from '../../store/actions';
+import {
+  fetchBasicGasEstimates,
+  setCustomGasLimit,
+  BASIC_ESTIMATE_STATES,
+  GAS_SOURCE,
+} from '../gas/gas.duck';
+import {
+  SET_BASIC_GAS_ESTIMATE_DATA,
+  BASIC_GAS_ESTIMATE_STATUS,
+  SET_ESTIMATE_SOURCE,
+} from '../gas/gas-action-constants';
+import {
+  QR_CODE_DETECTED,
+  SELECTED_ACCOUNT_CHANGED,
+  ACCOUNT_CHANGED,
+} from '../../store/actionConstants';
+import { calcTokenAmount } from '../../helpers/utils/token-util';
+import {
+  checkExistingAddresses,
+  isDefaultMetaMaskChain,
+  isOriginContractAddress,
+  isValidDomainName,
+} from '../../helpers/utils/util';
+import { getTokens, getUnapprovedTxs } from '../metamask/metamask';
+import { resetResolution } from '../ens';
+import {
+  isBurnAddress,
+  isValidHexAddress,
+} from '../../../shared/modules/hexstring-utils';
+
+// typedefs
+/**
+ * @typedef {import('@reduxjs/toolkit').PayloadAction} PayloadAction
+ */
+
+const name = 'send';
+
+/**
+ * The Stages that the send slice can be in
+ * 1. UNINITIALIZED - The send state is idle, and hasn't yet fetched required
+ *  data for gasPrice and gasLimit estimations, etc.
+ * 1. ADD_RECIPIENT - The user is selecting which address to send an asset to
+ * 2. DRAFT - The send form is shown for a transaction yet to be sent to the
+ *  Transaction Controller.
+ * 3. EDIT - The send form is shown for a transaction already submitted to the
+ *  Transaction Controller but not yet confirmed. This happens when a
+ *  confirmation is shown for a transaction and the 'edit' button in the header
+ *  is clicked.
+ */
+export const SEND_STAGES = {
+  UNINITIALIZED: 'UNINITIALIZED',
+  ADD_RECIPIENT: 'ADD_RECIPIENT',
+  DRAFT: 'DRAFT',
+  EDIT: 'EDIT',
+};
+
+/**
+ * The status that the send slice can be in is either
+ * 1. VALID - the transaction is valid and can be submitted
+ * 2. INVALID - the transaction is invalid and cannot be submitted
+ *
+ * A number of cases would result in an invalid form
+ * 1. The recipient is not yet defined
+ * 2. The amount + gasTotal is greater than the user's balance when sending
+ *  native currency
+ * 3. The gasTotal is greater than the user's *native* balance
+ * 4. The amount of sent asset is greater than the user's *asset* balance
+ * 5. Gas price estimates failed to load entirely
+ * 6. The gasLimit is less than 21000 (0x5208)
+ */
+export const SEND_STATUSES = {
```

Comment:
NIT: for all of these enums can the names be singular so that uses read e.g. `SEND_STATUS.VALID`. I find this more legible?

Reply:
There's a lot of precedent for pluralizing these and the constant references a group of STATUSES.

Take a look at constants/transactions, constants/alerts, constants/gas, (almost every constant file has plural). If we want to change this we should do it outside of this PR.